### PR TITLE
Add a force transfrom kvstore from v2 to v3

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: ref #3594

Problem Summary:
- After user running in mix mode and `kvstore` transform is very slow.

### What is changed and how it works?
- Force transform the `kvstore` from V2 to V3 when it restored.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 1]
```diff
@@ -170,6 +170,53 @@ PS::V1::PageStorage::Config getV1PSConfig(const PS::V2::PageStorage::Config & co
     return c;
 }
 
+void RegionPersister::forceTransformKVStoreV2toV3()
+{
+    assert(page_reader != nullptr);
+    assert(page_writer != nullptr);
+
+    Pages pages_transform = {};
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:178]: better to transform one by one for kvstore, or the memory consume could be high

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 2]
```diff
@@ -247,17 +286,36 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper,
...
+
+                forceTransformKVStoreV2toV3();
+                const auto & kvstore_remain_pages_after_transform = page_storage_v2->getNumberOfPages();
+                LOG_FMT_INFO(log, "Current kvstore translate to V3 finished. [ns_id={}] [done={}] [remain_before_translate_pages={}], [remain_after_translate_pages={}]", //
+                             ns_id,
+                             kvstore_remain_pages_after_transform == 0,
+                             kvstore_remain_pages,
+                             kvstore_remain_pages_after_transform);
+
+                if (kvstore_remain_pages_after_transform == 0)
```
Comment (lidezhu) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:310]: Under which case will `kvstore_remain_pages_after_transform` not be 0 after `forceTransformKVStoreV2toV3`?

Comment (jiaqizho) [reply dbms/src/Storages/Transaction/RegionPersister.cpp:310]: A logical error may cause that.

Will change it to `if (kvstore_remain_pages_after_transform != 0) throw`

**Review by lidezhu — APPROVED** [review]
LGTM

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 3]
```diff
@@ -170,6 +170,45 @@ PS::V1::PageStorage::Config getV1PSConfig(const PS::V2::PageStorage::Config & co
...
+
+        write_batch_transform.putPage(page.page_id, //
+                                      page_transform_entry.tag,
+                                      std::make_shared<ReadBufferFromMemory>(page.data.begin(),
+                                                                             page.data.size()),
+                                      page.data.size());
+
+        // Will rewrite into V3 one by one.
+        // The region data is big. It is not a good idea to combine pages.
+        page_writer->write(std::move(write_batch_transform), nullptr);
```
Comment (hehechen) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:200]: What happens if a region is written in V3 but not deleted in V2 ?

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 4]
```diff
@@ -247,18 +286,35 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper,
...
-                page_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, page_storage_v3);
-                page_reader = std::make_shared<PageReader>(run_mode, ns_id, /*storage_v2_*/ nullptr, page_storage_v3, global_context.getReadLimiter());
-            }
-            else
+            if (const auto & kvstore_remain_pages = page_storage_v2->getNumberOfPages(); kvstore_remain_pages != 0)
             {
                 page_writer = std::make_shared<PageWriter>(run_mode, page_storage_v2, page_storage_v3);
                 page_reader = std::make_shared<PageReader>(run_mode, ns_id, page_storage_v2, page_storage_v3, global_context.getReadLimiter());
+
+                forceTransformKVStoreV2toV3();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:294]:
```suggestion
                LOG_FMT_INFO(logger, "Current kvstore transform to V3 begin [pages_before_transform={}]", kvstore_remain_pages);
                forceTransformKVStoreV2toV3();
```

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 5]
```diff
@@ -247,18 +286,36 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper,
...
-            else
+            if (const auto & kvstore_remain_pages = page_storage_v2->getNumberOfPages(); kvstore_remain_pages != 0)
             {
                 page_writer = std::make_shared<PageWriter>(run_mode, page_storage_v2, page_storage_v3);
                 page_reader = std::make_shared<PageReader>(run_mode, ns_id, page_storage_v2, page_storage_v3, global_context.getReadLimiter());
+
+                LOG_FMT_INFO(logger, "Current kvstore transform to V3 begin [pages_before_transform={}]", kvstore_remain_pages);
+                forceTransformKVStoreV2toV3();
+                const auto & kvstore_remain_pages_after_transform = page_storage_v2->getNumberOfPages();
+                LOG_FMT_INFO(log, "Current kvstore transfrom to V3 finished. [ns_id={}] [done={}] [remain_before_translate_pages={}], [remain_after_translate_pages={}]", //
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:297]:
```suggestion
                LOG_FMT_INFO(log, "Current kvstore transfrom to V3 finished. [ns_id={}] [done={}] [pages_before_transform ={}], [pages_after_transform ={}]", //
```

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 6]
```diff
@@ -247,18 +286,36 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper,
...
-            else
+            if (const auto & kvstore_remain_pages = page_storage_v2->getNumberOfPages(); kvstore_remain_pages != 0)
             {
                 page_writer = std::make_shared<PageWriter>(run_mode, page_storage_v2, page_storage_v3);
                 page_reader = std::make_shared<PageReader>(run_mode, ns_id, page_storage_v2, page_storage_v3, global_context.getReadLimiter());
+
+                LOG_FMT_INFO(logger, "Current kvstore transform to V3 begin [pages_before_transform={}]", kvstore_remain_pages);
+                forceTransformKVStoreV2toV3();
+                const auto & kvstore_remain_pages_after_transform = page_storage_v2->getNumberOfPages();
+                LOG_FMT_INFO(log, "Current kvstore transfrom to V3 finished. [ns_id={}] [done={}] [pages_before_transform ={}], [pages_after_transform ={}]", //
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:297]:
```suggestion
                LOG_FMT_INFO(log, "Current kvstore transfrom to V3 finished. [ns_id={}] [done={}] [pages_before_transform ={}] [pages_after_transform ={}]", //
```

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 7]
```diff
@@ -247,18 +286,36 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper,
...
-            else
+            if (const auto & kvstore_remain_pages = page_storage_v2->getNumberOfPages(); kvstore_remain_pages != 0)
             {
                 page_writer = std::make_shared<PageWriter>(run_mode, page_storage_v2, page_storage_v3);
                 page_reader = std::make_shared<PageReader>(run_mode, ns_id, page_storage_v2, page_storage_v3, global_context.getReadLimiter());
+
+                LOG_FMT_INFO(logger, "Current kvstore transform to V3 begin [pages_before_transform={}]", kvstore_remain_pages);
+                forceTransformKVStoreV2toV3();
+                const auto & kvstore_remain_pages_after_transform = page_storage_v2->getNumberOfPages();
+                LOG_FMT_INFO(log, "Current kvstore transfrom to V3 finished. [ns_id={}] [done={}] [pages_before_transform ={}] [pages_after_transform ={}]", //
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:297]:
```suggestion
                LOG_FMT_INFO(log, "Current kvstore transfrom to V3 finished. [ns_id={}] [done={}] [pages_before_transform={}] [pages_after_transform={}]", //
```

File: dbms/src/Storages/Transaction/RegionPersister.cpp [hunk 8]
```diff
@@ -247,18 +286,36 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper,
...
-                page_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, page_storage_v3);
-                page_reader = std::make_shared<PageReader>(run_mode, ns_id, /*storage_v2_*/ nullptr, page_storage_v3, global_context.getReadLimiter());
-            }
-            else
+            if (const auto & kvstore_remain_pages = page_storage_v2->getNumberOfPages(); kvstore_remain_pages != 0)
             {
                 page_writer = std::make_shared<PageWriter>(run_mode, page_storage_v2, page_storage_v3);
                 page_reader = std::make_shared<PageReader>(run_mode, ns_id, page_storage_v2, page_storage_v3, global_context.getReadLimiter());
+
+                LOG_FMT_INFO(logger, "Current kvstore transform to V3 begin [pages_before_transform={}]", kvstore_remain_pages);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RegionPersister.cpp:294]:
```suggestion
                LOG_FMT_INFO(log, "Current kvstore transform to V3 begin [pages_before_transform={}]", kvstore_remain_pages);
```

**Pull request merged**